File: lc_go/lcGo/array/countMaxOrSubsets.go
package main

func countMaxOrSubsets(nums []int) (ans int) {
	n := len(nums)
	totalOr := 0
	for _, x := range nums {
		totalOr |= x
	}
	var dfs func(int, int)

	dfs = func(idx, currentSubsets int) {
		if idx == n {
			if currentSubsets == totalOr {
				ans++
			}
			return
		}
		dfs(idx+1, currentSubsets)
		dfs(idx+1, currentSubsets|nums[idx])
	}
	dfs(0, 0)
	return
}

func main() {

}

File: lc_cpp/lcCpp/backtrack/countMaxOrSubsets.cpp
#include <vector>

using namespace std;
class count_max_or_subsets {
public:
    int countMaxOrSubsets(vector<int>& nums) {
        int ans = 0;
        int total_or = reduce(nums.begin(), nums.end(), 0, bit_or());
        auto dfs = [&](this auto&& dfs, int i, int curr) -> void {
            if (i == nums.size()){
                if (curr == ans) {
                    ans++;
                }
                return ;
            }
            dfs(i + 1, curr);
            dfs(i + 1, curr | nums[i]);
        };

        dfs(0, 0);
        return ans;
    }
};

int main(){
    count_max_or_subsets solution;
    vector<int> nums = {3, 1};
    auto ans = solution.countMaxOrSubsets(nums);
    return 0;
}